### PR TITLE
[TextField] Add the `event` argument to the `onBlur` prop

### LINKED ISSUE #6115: Issue getting relatedTarget from onBlur via TextField
## Issue summary

On `shopify/web`, we have a DatePickerTextField component that relies on Polaris' TextField react component. There is [code that listens for the blur event](https://github.com/Shopify/web/blob/main/app/components/DatePickerTextField/DatePickerTextField.tsx#L170-L185) and destructures the event object passed in, expecting a `relatedTarget` property. According to git history, this code is around 2 years old. As of June 3, 2022, we've been seeing an error in `web` where `relatedTarget` cannot be destructured as no event object is passed in from the TextField's onBlur event. View an [example error on bugsnag](https://app.bugsnag.com/shopify/shopify-web-client/errors/629a015aaa0caf00090c2970?event_id=62a6e41c00950037410c0000&i=sk&m=ef).

## Expected behavior

We should be able to destructure the `relatedTarget` from the event object passed in from the blurred field. This worked in the past, although Polaris [documentation](https://polaris.shopify.com/components/forms/text-field) and [typing](https://github.com/Shopify/polaris/blob/main/polaris-react/src/components/TextField/TextField.tsx#L164) indicate that onBlur does not pass any params via `onBlur`, so I'm unsure exactly how the event object was were passed previously.

## Actual behavior

As of June 3, 2022 the event object is not being passed. I'm not positive but wonder if it might be related to [this change](https://github.com/Shopify/polaris/pull/5957/files#diff-b3c67fa075f17d88f8c43be7df7456856ae03cb49912c84f23560ddaf30d4fa4R476-R483) in Polaris to the TextField react component.

@Tom-Bonnike had reached out to me via Slack thinking it was related to a [recent PR I did](https://github.com/Shopify/web/pull/62766) in `shopify/web` that touched the DatePickerTextField component, but I temporarily reverted my PR locally on spin and the issue was still present. Digging in with @mattdrose we found the situation as described above.

Blurring the DatePickerTextField in the browser outputs an error like this to the browser dev console:

```
DatePickerTextField.tsx:170 Uncaught TypeError: Cannot destructure property 'relatedTarget' of 'undefined' as it is undefined.
    at handleInputBlur (DatePickerTextField.tsx:170:5)
    at handleOnBlur (TextField.js:254:7)
    at HTMLUnknownElement.callCallback2 (react-dom.development.js:3945:14)
    at HTMLUnknownElement.__trace__ (bugsnag.js:2765:17)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994:16)
    at invokeGuardedCallback (react-dom.development.js:4056:31)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070:25)
    at executeDispatch (react-dom.development.js:8243:3)
    at processDispatchQueueItemsInOrder (react-dom.development.js:8275:7)
    at processDispatchQueue (react-dom.development.js:8288:5)
```

## Steps to reproduce the problem

I have a new spin instance that has the issue.

1. Go to the staff admin page to be logged into admin. (https://shop1.shopify.datepicker-bug.bradley-staples.us.spin.dev/admin/staff)
2. Go to a page where the problem exists, such as https://shop1.shopify.datepicker-bug.bradley-staples.us.spin.dev/admin/transfers/new
3. Open the browser dev console. Then click into the "Estimated arrival" field under "Shipment details", and then blur the field. The error should be in your dev console. In the stack trace I posted above under Actual Behavior, the second line points to the TextField react component in Polaris.

## Specifications

- Are you using the React components? (Y/N): Yes
- Polaris version number: 9.14.1
- Browser: Chrome. Version 102.0.5005.115 (Official Build) (arm64)
- Device:  MacBook Pro (14-inch, 2021)
- Operating System: MacOS Monterey 12.4

Or run `npx envinfo --system --binaries --browsers --npmPackages react,react-dom,@shopify/polaris` to provide specifications on your environment including version numbers, browser, device, and operating system.

Paste the results here (I ran the command from my spin instance)

```bash
  System:
    OS: Linux 5.10 Ubuntu 21.10 21.10 (Impish Indri)
    CPU: (32) x64 Intel(R) Xeon(R) CPU @ 2.80GHz
    Memory: 38.52 GB / 125.90 GB
    Container: Yes
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 16.15.1 - ~/.nvm/versions/node/v16.15.1/bin/node
    Yarn: 1.22.15 - /usr/local/bin/yarn
    npm: 8.11.0 - ~/.nvm/versions/node/v16.15.1/bin/npm
    Watchman: 4.9.0 - /usr/bin/watchman
  npmPackages:
    @shopify/polaris: 9.14.1 => 9.14.1
    react: ~17.0.2 => 17.0.2
    react-dom: ~17.0.2 => 17.0.2
```

### PR DESCRIPTION
**Pull request by chloerice**
### WHY are these changes introduced?

Fixes #6115 

### WHAT is this pull request doing?

Callbacks set on the `TextField` `onBlur` prop had to hack around the lack of support for the `event` argument passed implicitly by the `input`  prior to shipping #5957. This PR fixes regressions upstream in `web` by adding and passing the `event` argument to the `onBlur` prop so that consuming apps don't need to hack their callback's types when composing complex inputs like `DatePickerTextField`.

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}
```

</details>

### 🎩 checklist

- [ ] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [x] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

File: polaris-react/src/components/TextField/TextField.tsx [hunk 1]
```diff
@@ -159,9 +159,9 @@ interface NonMutuallyExclusiveProps {
   /** Callback fired when value is changed */
   onChange?(value: string, id: string): void;
   /** Callback fired when input is focused */
-  onFocus?: (event?: React.FocusEvent<HTMLElement>) => void;
-  /** Callback fired when focus is removed */
-  onBlur?(): void;
+  onFocus?: (event?: React.FocusEvent) => void;
```
Comment (melnikov-s) [inline polaris-react/src/components/TextField/TextField.tsx]: I know it's not technically part of this PR but wondering why the `onFocus` has a optional `event` in the parameters. Shouldn't it always be there?

Comment (chloerice) [reply polaris-react/src/components/TextField/TextField.tsx]: Huh, good call I'm not sure 🤔

**Review by iangermann — APPROVED** [review]
lgtm! Thanks for the quick fix @chloerice!

**Review by melnikov-s — APPROVED** [review]
🎉

**Pull request merged**